Add SublimeLinter-contrib-terraform

patrickrgaffney opened the pull request:
https://github.com/patrickrgaffney/SublimeLinter-contrib-terraform

kaste commented:
Hi @patrickrgaffney! Nice work, I put some comments:

1. try to import `LintMatch` like `from SublimeLinter.lint import Linter, LintMatch` so the module `linter` itself stays an implementation detail.

2. `name = 'terraform'`is actually optional because we take the class name by default. (`name` is more for fancy things like `c#` which you couldn't do otherwise.)

3. For the 'severity', just use `LintMatch(error_type=severity, ...)`. (The warning/error thing is a special feat which allows easier regexes for some linter output formats.)

4. `LintMatch(match=...)` is optional, and it doesn't do anything atm. (Basically it's here for backwards compat.)

5. The filename. This feat is very young, and we actually hope it works plug'n'play. We support absolute and relative to the working dir paths here. What's the output from the linter here actually? I.e. `/f/b/t.tf`, `b/t.tf`, and `./b/t.tf` *should* work out of the box. If you want the file name, active folder etc., use e.g. `self.context.get('folder')` etc.

6. `cmd` should probably be `('terraform', 'validate', '--json', '${file_on_disk}')` if it's a file-on-disk (saved file only) linter.

Hm, looks like a lot, but I also just added tips'n'tricks here. And yes, our docs do not reflect the changes from this whole year. 🙄 ... like `LintMatch`, `filename` handling...

patrickrgaffney commented:
Thanks for the tips! I made a few different updates and cut a new release:

1. Switched to importing `LintMatch` directly `from SublimeLinter.lint`.
2. Removed the `name` property override, as it correctly infers `terraform`.
3. Assigning the `severity` value directly to `error_type` now. It will always be either `"warning"` or `"error"`.
4. Stopped setting the `match` property on the `LintMatch` object.
5. The `validate` command outputs basenames only (i.e. `foo.tf`) because it operates on a directory (without recursion). I had somehow managed to run into an issue where errors in one file were being highlighted in all `.tf` files. I *thought* I had *fixed* it in [`6b86ed8`](https://github.com/patrickrgaffney/SublimeLinter-contrib-terraform/commit/6b86ed817aa900236a4dc6a3030d8ec10b5f8a56). 😅 I switched just putting the filename directly into `LintMatch.filename` and (perhaps in combination with the next item?) appears to have resolved that issue.
6. I added the `file_path` argument because `terraform validate` requires a directory.

I was mostly poking around in other `contrib-` linters and the SublimeLinter source to figure out the best way to do this. I'll start watching the main repository and will try to help out where I can! Thanks again for the pointers!

kaste commented:
LGTM! @braver can we publish?

@patrickrgaffney If it doesn't work out with the filename, please open an issue at SublimeLinter core. It's a new feature and it will grow with experience.
